### ISSUE TITLE
Startup error related to HttpClient changes

### ISSUE THREAD

**Issue by vinnybod** [rubicon-project/prebid-server-java#205]
Getting an error starting up on my local machine (Mac OS Mojave), starting with commit https://github.com/rubicon-project/prebid-server-java/commit/46da2b0d9a6956f3183b3f452ceb9afbdd4a48fc

It seems to be having an issue initializing the `CurrencyConversionService` bean. All unit tests are passing.

```
2018-11-21 08:50:32.626  INFO 14437 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@32ee6fee: startup date [Wed Nov 21 08:50:32 MST 2018]; root of context hierarchy
2018-11-21 08:50:34.560  INFO 14437 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'springConfiguration' of type [org.prebid.server.spring.config.SpringConfiguration$$EnhancerBySpringCGLIB$$3a3233e5] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-11-21 08:50:34.582  INFO 14437 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'conversionService' of type [org.springframework.core.convert.support.DefaultConversionService] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-11-21 08:50:38.438  WARN 14437 --- [-thread-checker] io.vertx.core.impl.BlockedThreadChecker  : Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 2431 ms, time limit is 2000
2018-11-21 08:50:39.441  WARN 14437 --- [-thread-checker] io.vertx.core.impl.BlockedThreadChecker  : Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 3435 ms, time limit is 2000
2018-11-21 08:50:40.445  WARN 14437 --- [-thread-checker] io.vertx.core.impl.BlockedThreadChecker  : Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 4439 ms, time limit is 2000
2018-11-21 08:50:41.012  WARN 14437 --- [           main] s.c.a.AnnotationConfigApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'webConfiguration.AdminServerConfiguration': Unsatisfied dependency expressed through field 'currencyRatesHandler'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'currencyRatesHandler' defined in class path resource [org/prebid/server/spring/config/WebConfiguration$AdminServerConfiguration.class]: Unsatisfied dependency expressed through method 'currencyRatesHandler' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'currencyConversionRates' defined in class path resource [org/prebid/server/spring/config/ServiceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.prebid.server.currency.CurrencyConversionService]: Factory method 'currencyConversionRates' threw exception; nested exception is java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
2018-11-21 08:50:41.211  INFO 14437 --- [           main] utoConfigurationReportLoggingInitializer : 

Error starting ApplicationContext. To display the auto-configuration report re-run your application with 'debug' enabled.
2018-11-21 08:50:41.224 ERROR 14437 --- [           main] o.s.boot.SpringApplication               : Application startup failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'webConfiguration.AdminServerConfiguration': Unsatisfied dependency expressed through field 'currencyRatesHandler'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'currencyRatesHandler' defined in class path resource [org/prebid/server/spring/config/WebConfiguration$AdminServerConfiguration.class]: Unsatisfied dependency expressed through method 'currencyRatesHandler' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'currencyConversionRates' defined in class path resource [org/prebid/server/spring/config/ServiceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.prebid.server.currency.CurrencyConversionService]: Factory method 'currencyConversionRates' threw exception; nested exception is java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:588)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:366)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1264)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:761)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1118)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1107)
	at org.prebid.server.Application.main(Application.java:11)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'currencyRatesHandler' defined in class path resource [org/prebid/server/spring/config/WebConfiguration$AdminServerConfiguration.class]: Unsatisfied dependency expressed through method 'currencyRatesHandler' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'currencyConversionRates' defined in class path resource [org/prebid/server/spring/config/ServiceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.prebid.server.currency.CurrencyConversionService]: Factory method 'currencyConversionRates' threw exception; nested exception is java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:467)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1138)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:585)
	... 26 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'currencyConversionRates' defined in class path resource [org/prebid/server/spring/config/ServiceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.prebid.server.currency.CurrencyConversionService]: Factory method 'currencyConversionRates' threw exception; nested exception is java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1138)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:835)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741)
	... 39 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.prebid.server.currency.CurrencyConversionService]: Factory method 'currencyConversionRates' threw exception; nested exception is java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
	... 52 common frames omitted
Caused by: java.lang.RuntimeException: Action has not completed within defined timeout 5000 ms
	at org.prebid.server.vertx.ContextRunner.runOnContext(ContextRunner.java:80)
	at org.prebid.server.vertx.ContextRunner.runOnServiceContext(ContextRunner.java:50)
	at org.prebid.server.spring.config.ServiceConfiguration.currencyConversionRates(ServiceConfiguration.java:322)
	at org.prebid.server.spring.config.ServiceConfiguration$$EnhancerBySpringCGLIB$$81cbb55.CGLIB$currencyConversionRates$1(<generated>)
	at org.prebid.server.spring.config.ServiceConfiguration$$EnhancerBySpringCGLIB$$81cbb55$$FastClassBySpringCGLIB$$e9e88d1e.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:358)
	at org.prebid.server.spring.config.ServiceConfiguration$$EnhancerBySpringCGLIB$$81cbb55.currencyConversionRates(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 53 common frames omitted
```

If I raise `vertx.init-timeout-ms` to `20000`:
```
io.vertx.core.VertxException: Thread blocked
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:213)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:372)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$2.getObject(AbstractBeanFactory.java:345)
	at org.springframework.context.support.SimpleThreadScope.get(SimpleThreadScope.java:72)
	at org.prebid.server.spring.config.VertxContextScope.get(VertxContextScope.java:24)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:340)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.aop.target.SimpleBeanTargetSource.getTarget(SimpleBeanTargetSource.java:35)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:192)
	at com.sun.proxy.$Proxy49.get(Unknown Source)
	at org.prebid.server.currency.CurrencyConversionService.populatesLatestCurrencyRates(CurrencyConversionService.java:76)
	at org.prebid.server.currency.CurrencyConversionService.initialize(CurrencyConversionService.java:59)
	at org.prebid.server.spring.config.ServiceConfiguration.lambda$currencyConversionRates$0(ServiceConfiguration.java:323)
	at org.prebid.server.spring.config.ServiceConfiguration$$Lambda$60/923013679.handle(Unknown Source)
	at org.prebid.server.vertx.ContextRunner.lambda$runOnContext$2(ContextRunner.java:70)
	at org.prebid.server.vertx.ContextRunner$$Lambda$63/280223635.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:339)
	at io.vertx.core.impl.ContextImpl$$Lambda$64/3392189.run(Unknown Source)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by schernysh** [comment]
Thanks for reaching out.

This problem is not reproducible in our environments so far (Mac/Linux laptops, Linux servers). However we saw this issue lately and the commit you are referring to intended to fix this issue and it fixed the problem for us indeed. Not completely as it turned out.

Do you think you can try to diagnose the root cause and find a solution? That would be very helpful as currently you are the only one having a reproducible environment and hence could verify that fix works in fact.

Meanwhile we'll look into this problem again to find alternative solution.

**Comment by vinnybod** [comment]
Thanks for the quick response. I will see if I can reproduce it on another machine. It does start up in a docker container, so there's a place to start. I'm out for the next few days so I'll update this thread next week.

**Comment by schernysh** [comment]
I've been able to reproduce this issue running PBS in docker, investigating.

**Comment by schernysh** [comment]
I've created a PR #210 that hopefully fixes this issue finally. I will appreciate if you try to run it to verify the solution once you're back again.

**Comment by vinnybod** [comment]
I can confirm that I am able to startup on my Mac and Linux environments with the new change. Thanks for figuring out the problem!

**Comment by schernysh** [comment]
Fixed by #210.
